platform/pulp/tasking/system.py

```diff
@@ -0,0 +1,474 @@
+    :type  name:            basestring
+    :param normal_shutdown: True if the worker shutdown normally, False otherwise. Defaults to
+                            False.
+    :type normal_shutdown:  bool
+    """
+    if not normal_shutdown:
+        msg = _('The worker named %(name)s is missing. Canceling the tasks in its queue.')
+        msg = msg % {'name': name}
+        _logger.error(msg)
+
+    # Cancel all of the tasks that were assigned to this worker's queue
+    for task_status in Worker.objects.tasks.filter(state__in=constants.CALL_INCOMPLETE_STATES):
+        cancel(task_status.id)
+
+    # Delete the worker document
+    Worker.objects.get(name=name).delete()
+
+    # If the worker is a resource manager, we also need to delete the associated lock
+    if name.startswith(RESOURCE_MANAGER_WORKER_NAME):
+        ResourceManagerLock.objects.filter(name=name).delete()
+
+
+@task(base=PulpTask)
+def _release_resource(task_id):
+    """
+    Do not queue this task yourself. It will be used automatically when your task is dispatched by
+    the _queue_reserved_task task.
+
+    When a resource-reserving task is complete, this method releases the resource by removing the
+    ReservedResource object by UUID.
+
+    :param task_id: The UUID of the task that requested the reservation
+    :type  task_id: basestring
+    """
+    try:
+        TaskStatus.objects.get(id=task_id, state=constants.CALL_RUNNING_STATE)
+    except TaskStatus.DoesNotExist:
+        pass
+    else:
+        new_task = Task()
+        msg = _('The task status %(task_id)s exited immediately for some reason. Marking as '
+                'errored. Check the logs for more details')
+        runtime_exception = RuntimeError(msg % {'task_id': task_id})
+
+        class MyEinfo(object):
+            traceback = None
+
+        new_task.on_failure(runtime_exception, task_id, (), {}, MyEinfo)
+
+    ReservedResource.objects.get(task__id=task_id).delete()
+
+
+class UserFacingPulpTask(PulpTask):
+    """
+    A Pulp Celery task which will be visible to the user through the tasking portion of the API.
+
+    This object provides two interfaces to dispatch tasks: :meth: `apply_async` and
+    :meth: `apply_async_with_reservation`.
+
+    The :meth: `apply_asyc` provides normal celery dispatches of the task to the 'celery' queue
+    which all workers subscribe to. The task is handled by the first available worker.
+
+    The :meth: `apply_async_with_reservation` dispatch interface will send the task through the
+    resource_manager queue and process and will be assigned a specific worker. A series of
+    reservations cause specific types of tasks to not be run concurrently. See the
+    :meth: `apply_async_with_reservation` documentation for more details.
+    """
+
+    # this tells celery to not automatically log tracebacks for these exceptions
+    throws = (PulpCodedException,)
+
+    def apply_async_with_reservation(self, resource_type, resource_id, *args, **kwargs):
```

Comment:
We should consider doing something like this but I don't plan to do it with this PR. Is there some kind of issue tracking that idea?